### PR TITLE
[FLINK-33875][runtime] Support slots wait mechanism at DeclarativeSlotPoolBridge side for Default Scheduler

### PR DESCRIPTION
**Pull request by RocMarshal**
## What is the purpose of the change

- Refactor FreeSlotInfoTracker to FreeSlotTracker for the better semantic
- Support slots wait mechanism at DeclarativeSlotPoolBridge side for Default Scheduler

## Brief change log

- Refactor FreeSlotInfoTracker to FreeSlotTracker for the better semantic
- Support slots wait mechanism at DeclarativeSlotPoolBridge side for Default Scheduler

## Verifying this change

This change added tests and can be verified as the corresponding core test classes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery:  DefaultScheduler
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java [hunk 1]
```diff
@@ -221,7 +288,10 @@ void newSlotsAreAvailable(Collection<? extends PhysicalSlot> newSlots) {
                     slot.getAllocationId(),
                     pendingRequest.getResourceProfile());
         }
+    }
 
+    private void fulfillMatchedSlots(
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java:293]: nit: all callers call `reserveMatchedFreeSlots` and `fulfillMatchedSlots` together, why do we need 2 methods?

As I understand, merge them into one method is enough: `reserveAndFulfillMatchedFreeSlots`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java [hunk 2]
```diff
@@ -202,10 +206,73 @@ protected void onReleaseTaskManager(ResourceCounter previouslyFulfilledRequireme
...
+                    requestSlotMatches,
+                    pendingRequests);
+        } else {
+            // For requestSlotMatches.size() > pendingRequests.size()
+            throw new IllegalStateException(
+                    "The number of matched slots is not equals to the pendingRequests.");
+        }
+    }
+
+    private void newSlotsAvailableForNonSlotBatchAllocatable(
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java:258]:
```suggestion
    private void newSlotsAvailableForDirectlyAllocatable(
```

how about updating the name to `newSlotsAvailableForDirectlyAllocatable`?

**Comment by RocMarshal** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/AbstractDeclarativeSlotPoolBridgeTest.java [hunk 3]
```diff
@@ -51,37 +52,53 @@ abstract class AbstractDeclarativeSlotPoolBridgeTest {
...
     }
 
     @Nonnull
     DeclarativeSlotPoolBridge createDeclarativeSlotPoolBridge(
             DeclarativeSlotPoolFactory declarativeSlotPoolFactory,
-            RequestSlotMatchingStrategy requestSlotMatchingStrategy,
-            Duration slotRequestMaxInterval,
-            ComponentMainThreadExecutor mainThreadExecutor) {
+            ComponentMainThreadExecutor componentMainThreadExecutor,
+            DeclarativeSlotPoolBridgeResourceDeclarationTest.RequirementListener
```
Comment (1996fanrui) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/AbstractDeclarativeSlotPoolBridgeTest.java:100]: Is it possible that we don't pass `RequirementListener` here?

Alternative solution is : All tests of DeclarativeSlotPoolBridgeResourceDeclarationTest could call `requirementListener#tryWaitSlotRequestIsDone` directly.

In general, the parent class doesn't depends on the class inside of children class.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java [hunk 4]
```diff
@@ -548,4 +610,13 @@ void increaseResourceRequirementsBy(ResourceCounter increment) {
...
     }
+
+    @VisibleForTesting
+    public void tryWaitSlotRequestIsDone() {
+        if (getDeclarativeSlotPool() instanceof DefaultDeclarativeSlotPool) {
+            final DefaultDeclarativeSlotPool slotPool =
+                    (DefaultDeclarativeSlotPool) getDeclarativeSlotPool();
+            slotPool.tryWaitSlotRequestIsDone();
+        }
+    }
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java:621]: If last comment makes sense, could we move ` void tryWaitSlotRequestIsDone()` method to `DeclarativeSlotPoolBridgeTest`?

Only `DeclarativeSlotPoolBridgeTest` calls `tryWaitSlotRequestIsDone()`.

**Review by 1996fanrui — COMMENTED** [review]
Thanks @RocMarshal for the update!

Overall the PR LGTM, I only left some comments for test code. Please take a look in your free time, thanks~

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeResourceDeclarationTest.java [hunk 5]
```diff
@@ -205,12 +223,36 @@ void testRequirementsDecreasedOnSlotAllocationFailure() throws Exception {
                 .isZero();
     }
 
-    private static final class RequirementListener {
+    /** Requirement listener for testing. */
+    static final class RequirementListener {
```
Comment (1996fanrui) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeResourceDeclarationTest.java:227]: After refactor, the original private may be enough, right?

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeTest.java [hunk 6]
```diff
@@ -252,4 +261,76 @@ void testIfJobIsRestartingAllOfferedSlotsWillBeRegistered() throws Exception {
...
+            if (slotBatchAllocatable) {
+                slotAllocationFuture1.join();
+                slotAllocationFuture2.join();
+            } else {
+                slotAllocationFuture2.join();
+            }
+        }
+    }
+
+    void tryWaitSlotRequestIsDone(DeclarativeSlotPoolBridge declarativeSlotPoolBridge) {
```
Comment (1996fanrui) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridgeTest.java:312]: Could it be private?

**Review by 1996fanrui — CHANGES_REQUESTED** [review]

**Comment by RocMarshal** [comment]
HI, @1996fanrui Thank you very much for the quick review. I made some change based on your comments. PTAL if you had the free time~ :)

**Review by 1996fanrui — APPROVED** [review]
Thanks @RocMarshal for the update!

LGTM assuming the CI is green.

**Pull request merged**